pkg/controller/networkpolicy/clusternetworkpolicy.go

```diff
@@ -286,89 +288,140 @@ func (n *NetworkPolicyController) processClusterNetworkPolicy(cnp *crdv1alpha1.C
-						klog.V(4).Infof("Adding a new per-namespace rule with appliedTo %v for rule %d of %s", clusterAppliedToAffectedNS[i], idx, cnp.Name)
-						addRule(n.toNamespacedPeerForCRD(perNSPeers, clusterAppliedToAffectedNS[i]), direction, []string{atgForNamespace[i]})
+						klog.V(4).InfoS("Adding a new per-namespace rule with appliedTo a ClusterNetworkPolicy rule", "appliedToAffectedNS", clusterAppliedToAffectedNS[i], "ruleIndex", idx, "clusterNetworkPolicy", acnp.Name)
+						addRule(n.toNamespacedPeerForCRD(perNSPeers, clusterAppliedToAffectedNS[i]), services, direction, []string{atgForNamespace[i]})
 					}
 				} else {
 					// Create a rule for each affected Namespace of appliedTo at rule level
-					for _, at := range cnpRule.AppliedTo {
+					for _, at := range acnpRule.AppliedTo {
 						affectedNS, selectors := n.getAffectedNamespacesForAppliedTo(at)
 						affectedNamespaceSelectors = append(affectedNamespaceSelectors, selectors...)
 						for _, ns := range affectedNS {
 							atg := n.createAppliedToGroup(ns, at.PodSelector, nil, at.ExternalEntitySelector)
 							atgNamesSet.Insert(atg)
-							klog.V(4).Infof("Adding a new per-namespace rule with appliedTo %v for rule %d of %s", atg, idx, cnp.Name)
-							addRule(n.toNamespacedPeerForCRD(perNSPeers, ns), direction, []string{atg})
+							klog.V(4).InfoS("Adding a new per-namespace rule with appliedTo for a ClusterNetworkPolicy rule", "appliedToGroup", atg, "ruleIndex", idx, "clusterNetworkPolicy", acnp.Name)
+							addRule(n.toNamespacedPeerForCRD(perNSPeers, ns), services, direction, []string{atg})
 						}
 					}
 				}
 			}
+		} else {
+			// Handle toService rules.
+			ruleAppliedTos := acnpRule.AppliedTo
+			if appliedToPerRule && len(acnp.Spec.AppliedTo) > 0 {
+				ruleAppliedTos = acnp.Spec.AppliedTo
+			}
+			ruleATGNames := n.processClusterAppliedTo(ruleAppliedTos, atgNamesSet)
+			// Because each ServiceReference refers to its own combination of ports and
+			// Pods/IPBlocks，we need to add an individual rule for each ServiceReference.
+			for _, eachService := range acnpRule.ToServices {
+				svc, err := n.serviceLister.Services(eachService.Namespace).Get(eachService.Name)
+				if err != nil {
+					klog.V(2).InfoS("Service referred in `toServices` doesn't exist", "serviceNamespace", eachService.Namespace, "serviceName", eachService.Name)
+					continue
+				}
+				if svc.Spec.Type != v1.ServiceTypeClusterIP {
+					if svc.Spec.Type == v1.ServiceTypeNodePort {
+						klog.V(2).InfoS("Processing NodePort Service: still install rules matching the Endpoints IP+Ports, while no rule matching NodePort will be installed", "serviceNamespace", eachService.Namespace, "serviceName", eachService.Name, "serviceType", svc.Spec.Type)
+					} else {
+						klog.V(2).InfoS("Received unsupported ServiceType Service. Skip this Service", "serviceNamespace", eachService.Namespace, "serviceName", eachService.Name, "serviceType", svc.Spec.Type)
+					}
+				}
```

Comment:
Because of the verbosity these logs won't show by default. We also log the same thing every time rules are processed, which is not good. We only want to log something once per Service IMO.

Reply:
We can check if `internalNetworkPolicyStore` has this NP. But this will also skip this logging when a Service changes its type to an unsupported type. Also, I think it will be too complicated to check whether ServiceType has been updated here. What does this approach sound to you?
And we have other logs that will log the same thing every time rules are processed. So I guess it's not that bad if we log the same thing?